Mathlib/Data/Finset/Antidiagonal.lean

```diff
@@ -0,0 +1,183 @@
+class HasMulAntidiagonal (μ : Type*) [Monoid μ] where
+  /-- The antidiagonal function -/
+  antidiagonal : μ → Finset (μ × μ)
+  /-- A pair belongs to `antidiagonal n` iff the product of its components is equal to `n` -/
+  mem_antidiagonal : ∀ (n : μ) (a : μ × μ), a ∈ antidiagonal n ↔ a.fst * a.snd = n
+
+/-- The class of additive monoids with an antidiagonal -/
+class HasAntidiagonal (μ : Type*) [AddMonoid μ] where
+  /-- The antidiagonal function -/
+  antidiagonal : μ → Finset (μ × μ)
+  /-- A pair belongs to `antidiagonal n` iff the sum of its components is equal to `n` -/
+  mem_antidiagonal : ∀ (n : μ) (a : μ × μ), a ∈ antidiagonal n ↔ a.fst + a.snd = n
+
+variable {μ : Type _}
+  [CanonicallyOrderedAddCommMonoid μ]
+  [LocallyFiniteOrder μ] [DecidableEq μ]
+
+variable {ι : Type _} [DecidableEq ι] [DecidableEq (ι → μ)]
+
+/-- In a canonically ordered add monoid, the antidiagonal can be construct by filtering.
+
+Note that this is not an instance, as for some times a more efficient algorithm is available. -/
+abbrev antidiagonalOfLocallyFinite : HasAntidiagonal μ where
+  antidiagonal n := Finset.filter (fun uv => uv.fst + uv.snd = n) (Finset.product (Iic n) (Iic n))
+  mem_antidiagonal n a := by
+    simp only [Prod.forall, mem_filter, and_iff_right_iff_imp]
+    intro h; rw [← h]
+    erw [mem_product, mem_Iic, mem_Iic]
+    exact ⟨le_self_add, le_add_self⟩
+
+/- These functions apply to (ι →₀ ℕ), more generally to (ι →₀ μ) under the additional assumption `OrderedSub μ` that make it a canonically ordered add monoid
+In fact, we just need an AddMonoid with a compatible order,
+finite Iic, such that if a + b = n, then a, b ≤ n,
+and any other bound would be OK.
+
+  What follows is an analogous definition for ι → μ, with additional conditions on the support
+-/
+
+/-- The Finset of functions `ι → μ` whose support is contained in `s : Finset ι`
+  and whose sum is `n` -/
+def piAntidiagonal (s : Finset ι) (n : μ) : Finset (ι → μ) :=
+  Finset.filter (fun f => s.sum f = n)
+    ((s.pi fun _ => Iic n).map
+      ⟨fun f i => if h : i ∈ s then f i h else 0,
+        fun f g h => by ext i hi; simpa only [dif_pos hi] using congr_fun h i⟩)
+
+@[simp]
+theorem mem_piAntidiagonal (s : Finset ι) (n : μ) (f : ι → μ) :
+    f ∈ piAntidiagonal s n ↔ s.sum f = n ∧ ∀ (i) (_ : i ∉ s), f i = 0 := by
+  rw [piAntidiagonal, mem_filter, and_comm, and_congr_right]
+  intro h
+  simp only [mem_map, exists_prop, Function.Embedding.coeFn_mk, mem_pi]
+  constructor
+  · rintro ⟨_, _, rfl⟩ _ hi
+    dsimp only
+    rw [dif_neg hi]
+  · intro hf
+    refine' ⟨fun i _ => f i, fun i hi => _, _⟩
+    · rw [mem_Iic, ← h]
+      apply single_le_sum _ hi
+      simp only [zero_le, implies_true]
+    · ext x
+      dsimp only
+      rw [dite_eq_ite, ite_eq_left_iff, eq_comm]
+      exact hf x
+
+theorem piAntidiagonal_equiv_antidiagonal (n : μ) :
+    Equiv.finsetCongr (Equiv.boolArrowEquivProd _) (piAntidiagonal univ n) = antidiagonalOfLocallyFinite.antidiagonal n := by
```

Comment:
Here you should be taking `[HasAntidiagonal μ]` instead of relying on a specific instance.

Reply:
yep, and at two other places below as well!